pennylane/_qubit_device.py

```diff
@@ -1014,13 +1015,29 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
                 if op.name not in ("QubitStateVector", "BasisState", "Snapshot"):
                     expanded_ops.append(op)
 
-        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+        trainable_params = []
+        for k in tape.trainable_params:
+            # pylint: disable=protected-access
+            if hasattr(tape._par_info[k]["op"], "return_type"):
+                warnings.warn(
```

Comment:
Do you think it is better to have a warning rather than raising an error?

Reply:
Good question, I must admit, I'm not sure which to prefer here. It more so felt like something that "we can fix" for the user while also informing them, that's why I chose a warning.

Another more prominent advantage of a warning is that user code will not start to break right away. If we wanted to add an error here, it may still be advisable to start off with a warning first and then make it an error the release after.

---

Warning:
* `+`: the execution doesn't stop, useful when `requires_grad=True` is set by default;
* `-`: the user may ignore the warning and think that parameters to `qml.Hermitian` were trained.

Error:
* `+`: The user is notified for sure;
* `-`: A more complex workflow would be interrupted due to the error.

Reply:
Good point.

One occurrence we have for another warning is for the parameter-shift transform here:
https://github.com/PennyLaneAI/pennylane/blob/55381d38adab411d5a44027260a9626abf6e464d/pennylane/gradients/parameter_shift.py#L623

Reply:
Therefore I think we should continue with a warning here 👍